File: src/main/java/com/monovore/hunger/Communicator.java
package com.monovore.hunger;

import org.apache.kafka.clients.ClientRequest;
import org.apache.kafka.clients.KafkaClient;
import org.apache.kafka.common.Node;
import org.apache.kafka.common.errors.BrokerNotAvailableException;
import org.apache.kafka.common.errors.DisconnectException;
import org.apache.kafka.common.protocol.ApiKeys;
import org.apache.kafka.common.protocol.types.Struct;
import org.apache.kafka.common.requests.*;
import org.apache.kafka.common.utils.Time;

import java.io.Closeable;
import java.io.IOException;
import java.util.Optional;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.LinkedBlockingQueue;

public class Communicator implements Closeable {

    private final KafkaClient kafka;
    private final BlockingQueue<Outbound> outbound;
    private final Time time;
    private final Thread thread;

    private volatile boolean running = true;

    private static final long POLL_TIMEOUT_MS = 1000L;

    public Communicator(KafkaClient kafka, Time time) {
        this.kafka = kafka;
        this.outbound = new LinkedBlockingQueue<>();
        this.time = time;
        this.thread = new Thread(this::run);
        this.thread.setDaemon(true);
        this.thread.start();
    }

    public ApiClient anyone() {
        return new ApiClient((key, request) -> send(Optional.empty(), key, request));
    }

    public ApiClient node(Node node) {
        return new ApiClient(((key, request) -> send(Optional.of(node), key, request)));
    }

    public CompletableFuture<GroupClient<Void>> group(GroupClient.Meta meta) {
        return Futures.lifting(GroupCoordinatorResponse::errorCode, this.anyone().groupCoordinator(new GroupCoordinatorRequest(meta.groupId)))
                .thenApply(groupCoordinatorResponse -> new GroupClient<Void>(this.node(groupCoordinatorResponse.node()), meta, null));
    }

    public CompletableFuture<ApiClient> coordinator(String groupID) {
        return anyone().groupCoordinator(new GroupCoordinatorRequest(groupID))
                .thenCompose(Futures.liftError(GroupCoordinatorResponse::errorCode))
                .thenApply(response -> node(response.node()));
    }

    private CompletableFuture<Struct> send(Optional<Node> destination, ApiKeys key, Struct request) {
        CompletableFuture<Struct> responseFuture = new CompletableFuture<>();

        if (!running) {
          responseFuture.completeExceptionally(new IllegalStateException("Client has been closed!"));
          return responseFuture;
        }

        try {
            this.outbound.put(new Outbound(destination, key, request, responseFuture));
        } catch (InterruptedException e) {
            responseFuture.completeExceptionally(e);
            Thread.currentThread().interrupt();
        }
        this.kafka.wakeup();
        return responseFuture;
    }

    @Override
    public void close() throws IOException {
        running = false;
        kafka.close();
    }

    private void run() {
        Outbound next;
        while(running && !Thread.interrupted()) {
            long now = time.milliseconds();
            while ((next = outbound.poll()) != null) {
                final Outbound message = next;
                Node destination =
                        message.destination.orElseGet(() -> kafka.leastLoadedNode(now));

                if (destination == null) {
                    message.responseFuture.completeExceptionally(new BrokerNotAvailableException("No brokers available!"));
                    continue;
                }

                if (!kafka.ready(destination, now)) kafka.poll(POLL_TIMEOUT_MS, now);

                if (!kafka.isReady(destination, now)) {
                    message.responseFuture.completeExceptionally(new BrokerNotAvailableException("Connection not ready to node: " + destination.idString()));
                    continue;
                }

                RequestHeader header = kafka.nextRequestHeader(message.key);
                RequestSend send = new RequestSend(destination.idString(), header, message.request);
                ClientRequest request = new ClientRequest(now, true, send, response -> {
                    if (response.wasDisconnected()) {
                        message.responseFuture.completeExceptionally(new DisconnectException("Disconnected from node: " + destination.idString()));
                    } else {
                        Struct body = response.responseBody();
                        message.responseFuture.complete(body);
                    }
                });
                kafka.send(request, now);
            }
            // TODO: deal with exceptions here
            kafka.poll(POLL_TIMEOUT_MS, now);
        }
    }

    private static class Outbound {
        final Optional<Node> destination;
        final ApiKeys key;
        final Struct request;
        final CompletableFuture<Struct> responseFuture;

        Outbound(Optional<Node> destination, ApiKeys key, Struct request, CompletableFuture<Struct> responseFuture) {
            this.destination = destination;
            this.key = key;
            this.request = request;
            this.responseFuture = responseFuture;
        }
    }
}

File: src/main/scala/com/monovore/hunger/Communicator.scala
package com.monovore.hunger

import java.io.Closeable
import java.util
import java.util.concurrent.LinkedBlockingQueue

import cats.effect.Async
import org.apache.kafka.clients.{ClientRequest, ClientResponse, KafkaClient}
import org.apache.kafka.clients.consumer.internals.NoAvailableBrokersException
import org.apache.kafka.common.Node
import org.apache.kafka.common.errors.{BrokerNotAvailableException, DisconnectException}
import org.apache.kafka.common.protocol.ApiKeys
import org.apache.kafka.common.protocol.types.Struct
import org.apache.kafka.common.requests.RequestSend
import org.apache.kafka.common.utils.Time

import scala.annotation.tailrec

import scala.collection.JavaConverters._

object Communicator {

  private[Communicator] val PollTimeoutMs = 1000L

  case class Outbound(destination: Option[Node], key: ApiKeys, request: Struct, callback: Either[Throwable, Struct] => Unit)
}

/**
  * An event loop wrapped around the standard Kafka network client.
  * 
  */
class Communicator(val kafka: KafkaClient, val time: Time) extends Runnable with Closeable {

  private[this] val outbound: LinkedBlockingQueue[Communicator.Outbound] = new LinkedBlockingQueue[Communicator.Outbound]

  private[this] var running: Boolean = true

  def anyone[F[_]: Async] = new ApiClient[F]((key: ApiKeys, request: Struct) => send(None, key, request))

  def node[F[_]: Async](node: Node) = new ApiClient[F]((key: ApiKeys, request: Struct) => send(Some(node), key, request))

  private def send[F[_]](
    destination: Option[Node],
    key: ApiKeys,
    request: Struct
  )(implicit
    A: Async[F]
  ) = A.async[Struct] { callback =>
    if (!running) {
      callback(Left(new IllegalStateException("Client has been closed!")))
    }
    try
      this.outbound.put(Communicator.Outbound(destination, key, request, callback))
    catch {
      case e: InterruptedException =>
        callback(Left(e))
        Thread.currentThread.interrupt()
    }
    this.kafka.wakeup()
  }

  override def close(): Unit = {
    running = false
    kafka.close()
  }

  def run() = {

    @tailrec
    def drain(now: Long): Unit = {
      val arrayList = new util.ArrayList[Communicator.Outbound]()
      val numberToSend = outbound.drainTo(arrayList)

      for (message <- arrayList.iterator.asScala) {
        val destination = message.destination.getOrElse(kafka.leastLoadedNode(now))
        if (destination == null) Left(new BrokerNotAvailableException("No brokers available!"))
        if (!kafka.isReady(destination, now)) {
          message.callback(Left(new BrokerNotAvailableException("Connection not ready to node: " + destination.idString)))
            continue //todo: continue is not supported
        }
        val header = kafka.nextRequestHeader(message.key)
        val send = new RequestSend(destination.idString, header, message.request)
        val request = new ClientRequest(now, true, send, (response: ClientResponse) => {
          def foo(response: ClientResponse) = if (response.wasDisconnected) message.callback(Left((new DisconnectException("Disconnected from node: " + destination.idString))
          else {
            val body = response.responseBody
            message.responseFuture.complete(body)
          }

          foo(response)
        })
        kafka.send(request, now)
      }

      Option(outbound.poll()) match {
        case None => ()
        case Some(message) =>
          message.destination.orElse(Option(kafka.leastLoadedNode(now))) match {
            case None =>
              message.callback(Left(new NoAvailableBrokersException()))
            case Some(destination) =>
          }
      }
    }

    while (running) {
      val now = time.milliseconds
      var next: Communicator.Outbound = null
      while ((next = outbound.poll()) != null) {
      val message = next
        val destination = message.destination.getOrElse(kafka.leastLoadedNode(now))
        if (destination == null) {
          message.callback(Left((new BrokerNotAvailableException("No brokers available!"))
          continue //todo: continue is not supported
        }
        if (!kafka.ready(destination, now)) kafka.poll(Communicator.PollTimeoutMs, now)
        if (!kafka.isReady(destination, now)) {
          message.callback(Left((new BrokerNotAvailableException("Connection not ready to node: " + destination.idString))
          continue //todo: continue is not supported
        }
        val header = kafka.nextRequestHeader(message.key)
        val send = new RequestSend(destination.idString, header, message.request)
        val request = new ClientRequest(now, true, send, (response: ClientResponse) => {
          def foo(response: ClientResponse) = if (response.wasDisconnected) message.callback(Left((new DisconnectException("Disconnected from node: " + destination.idString))
          else {
            val body = response.responseBody
            message.responseFuture.complete(body)
          }

          foo(response)
        })
        kafka.send(request, now)
      }
      // TODO: deal with exceptions here
      kafka.poll(Communicator.PollTimeoutMs, now)
    }

    ???
  }
}
